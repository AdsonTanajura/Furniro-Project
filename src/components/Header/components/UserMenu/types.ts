export type UserMenuProps = {
  setIsCartOpen: (value: boolean) => void;
};
