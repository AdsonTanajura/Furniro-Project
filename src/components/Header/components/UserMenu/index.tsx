import {
  CartButton,
  CartIcon,
  StyledSignInButton,
  UserMenuContainer,
} from './styles';
import { SignedIn, SignedOut, UserButton } from '@clerk/clerk-react';
import userIcon from '../../../../assets/User.svg';
import UserMenuProps from './interfaces/UserMenuProps';

const UserMenu = ({ setIsCartOpen }: UserMenuProps) => {
  const handleClickCartButton = () => {
    setIsCartOpen(true);
  };
  return (
    <UserMenuContainer>
      <SignedOut>
        <StyledSignInButton>
          <img src={userIcon} alt="user avata" />
        </StyledSignInButton>
      </SignedOut>
      <SignedIn>
        <UserButton />
      </SignedIn>
      <CartButton aria-label="cart" onClick={handleClickCartButton}>
        <CartIcon data-testid="cart-icon" />
      </CartButton>
    </UserMenuContainer>
  );
};

export default UserMenu;
