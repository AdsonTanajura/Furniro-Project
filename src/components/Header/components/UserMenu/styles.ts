import { SignInButton } from '@clerk/clerk-react';
import { MdOutlineShoppingCart } from 'react-icons/md';
import styled from 'styled-components';

export const UserMenuContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
  gap: 8px;
`;

export const StyledSignInButton = styled(SignInButton)`
  cursor: pointer;
  padding: 8px;
  transition: border-bottom 0.2s;

  &:hover {
    border-bottom: 2px #000000 solid;
  }
`;

export const CartButton = styled.button`
  display: flex;
  background-color: transparent;
  border: none;
  cursor: pointer;
  padding: 4px;

  transition: border-bottom 0.2s;
  &:hover {
    border-bottom: 2px #000000 solid;
  }
`;

export const CartIcon = styled(MdOutlineShoppingCart)`
  width: 28px;
  height: 28px;
`;
