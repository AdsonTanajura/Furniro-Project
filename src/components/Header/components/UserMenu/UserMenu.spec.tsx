import { render, waitFor } from '@testing-library/react';
import { ClerkProvider } from '@clerk/clerk-react';
import userEvent from '@testing-library/user-event';
import UserMenu from '.';

jest.mock('@clerk/clerk-react');

const mockSetIsCartOpen = jest.fn();

describe('<UserMenu />', () => {
  it('should render the UserMenuContainer', () => {
    const { container } = render(
      <ClerkProvider publishableKey="test-clerk-key">
        <UserMenu setIsCartOpen={mockSetIsCartOpen} />
      </ClerkProvider>
    );
    expect(container.firstChild).toBeInTheDocument();
    expect(container.firstChild).toBeVisible();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('function should return SetIsCartOpen when the cart button is clicked', async () => {
    const { getByTestId, getByRole, getAllByAltText, getAllByRole, container } =
      render(
        <ClerkProvider publishableKey="test-clerk-key">
          <UserMenu setIsCartOpen={mockSetIsCartOpen} />
        </ClerkProvider>
      );
    const buttons = getAllByRole('button');
    const userAvatars = getAllByAltText('user avata');
    const cartButton = getByRole('button', { name: /cart/i });
    const cartIcon = getByTestId('cart-icon');

    expect(mockSetIsCartOpen).not.toHaveBeenCalled();
    expect(buttons.length).toBeLessThanOrEqual(3);
    expect(buttons.length).toBeGreaterThan(0);

    expect(userAvatars.length).toBeLessThanOrEqual(1);

    expect(cartButton).toBeInTheDocument();
    expect(cartButton).not.toBeDisabled();

    expect(cartIcon).toBeInTheDocument();

    await userEvent.click(cartButton);
    await waitFor(() => expect(mockSetIsCartOpen).toHaveBeenCalled());
    await waitFor(() => expect(mockSetIsCartOpen).toHaveBeenCalledWith(true));

    expect(container.firstChild).toMatchSnapshot();
  });
});
