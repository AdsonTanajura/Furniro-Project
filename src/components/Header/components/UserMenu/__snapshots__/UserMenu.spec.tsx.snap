// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserMenu /> function should return SetIsCartOpen when the cart button is clicked 1`] = `
<div
  class="sc-VHjGu dNpwMk"
>
  <button
    data-testid="clerk-sign-in-button"
  >
    <img
      alt="user avata"
      src="test-file-stub"
    />
  </button>
  <button />
  <button
    aria-label="cart"
    class="sc-dQkurY jQqHmu"
  >
    <svg
      class="sc-glJfXD kfDKCa"
      data-testid="cart-icon"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M15.55 13c.75 0 1.41-.41 1.75-1.03l3.58-6.49A.996.996 0 0 0 20.01 4H5.21l-.94-2H1v2h2l3.6 7.59-1.35 2.44C4.52 15.37 5.48 17 7 17h12v-2H7l1.1-2h7.45zM6.16 6h12.15l-2.76 5H8.53L6.16 6zM7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
  </button>
</div>
`;

exports[`<UserMenu /> should render the UserMenuContainer 1`] = `
<div
  class="sc-VHjGu dNpwMk"
>
  <button
    data-testid="clerk-sign-in-button"
  >
    <img
      alt="user avata"
      src="test-file-stub"
    />
  </button>
  <button />
  <button
    aria-label="cart"
    class="sc-dQkurY jQqHmu"
  >
    <svg
      class="sc-glJfXD kfDKCa"
      data-testid="cart-icon"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M15.55 13c.75 0 1.41-.41 1.75-1.03l3.58-6.49A.996.996 0 0 0 20.01 4H5.21l-.94-2H1v2h2l3.6 7.59-1.35 2.44C4.52 15.37 5.48 17 7 17h12v-2H7l1.1-2h7.45zM6.16 6h12.15l-2.76 5H8.53L6.16 6zM7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
  </button>
</div>
`;
